### ISSUE TITLE
Code cleanup and refactoring, increasing tests

### ISSUE THREAD

**Issue by Divy123** [publiclab/image-sequencer#1127]
This issue is a general discussion issue where we discuss upon code cleanup, refactoring and testing:
 - [ ] Correct insert step functionality
 - [ ] Increase number of tests to minimize  code breaks 
 - [ ] Deciding upon a code-editor recommendation for better debugging
 - [ ] Refactoring parts of code to make it more modular
 - [ ] moving tests from tape and jasmine to jest
 - [ ] moving from grunt to webpack

**Comment by Divy123** [comment]
@aashna27 @tech4GT @jywarren we can discuss important parts of code cleaning and refactoring here,
Linking already existing issues now.

**Comment by Divy123** [comment]
These are some of the issues that should be closed on a high priority.

Also some issues like moving to webpack and jest require to discuss upon.

I think @aashna27  and me can divide this and continue to improve upon.
Also I will be discussing to refactor parts of code as I will be going through solving some of the avbove issues.

I am working on insert step functionality right. 
Also what do you all think on moving to webpack and jest?

**Comment by Divy123** [comment]
Also, as I and @tech4GT discussed on a call if we can have intelliJ as our recommended editor to get the debugging done at its best as it provides some of the best tools through various extensions, it has got and have a wiki post on publiclab.org ??

**Comment by aashna27** [comment]
Yeah cool i currently have one pr open after which i can start working on
these.

 I would like to share that i tried working on ui testing using jasmine
jquery, it is similar to jest but i was stuck with the spyOnEvent(), for
dom testing. It did not work. I couldn't figure out how to proceed and as I
was running out of time I had to move on from it. Using fixtures for dom
testing was complex as a lot of our html is generated in Javascript.

**Comment by Divy123** [comment]
So yeah I think we have another reason for Jest here, the Facebook community behind the library is very active and we can get over things like this very easily .

**Comment by Divy123** [comment]
@HarshKhandeparkar I am seeing this error when clicking on insert step button. Needs urgent attention.
![Screenshot from 2019-06-25 15-45-39](https://user-images.githubusercontent.com/31225007/60090518-53cb5580-9760-11e9-920f-4fcbb0a8f226.png)

**Comment by Divy123** [comment]
Also after refreshing the page, the url is correct but the steps inside url are not getting loaded.

**Comment by Divy123** [comment]
Ok, I figured it out and what's happening is inside, defaultHtmlStepUi.js:
```js
if (step.linkElements[index].contains(imgthumbnail))
        step.linkElements[index].href = step.imgElement.src;
    }
```
at the time of checking the condition, step.linkElements[index] is unavailable, maybe some async stuff here.

@HarshKhandeparkar can you figure out please what part of code is making this happen?

**Comment by Divy123** [comment]
The insertStep is broken very badly, @publiclab/is-reviewers if you can help me here.

**Comment by tech4GT** [comment]
@jywarren I think we should revert back the last commit!

**Comment by jywarren** [comment]
OK - can someone revert it and open a PR for it? Thanks. Don't worry, folks, we'll resolve this and get it re-merged before long. 

Let's try to document the errors and write tests against them so we will know when we can re-merge, please?

**Comment by tech4GT** [comment]
@Divy123 Can you please open the pr ??

**Comment by Divy123** [comment]
I have reverted some files causing the major errors in one of my PR.
Here is the link (merged already): https://github.com/publiclab/image-sequencer/pull/1131

Also for the permanent fix, Harsh has opened a PR:
https://github.com/publiclab/image-sequencer/pull/1132

**Comment by HarshKhandeparkar** [comment]
@Divy123 my scopeQuery fix also fixes insert step along the way. I won't recommend intelliJ as it is a huge editor(size). I would rather recommend vscode.

Vscode already has intelisense baked in. They even added a full fledged debugger recently.

**Comment by Divy123** [comment]
@HarshKhandeparkar request you to send a gift for completely working functionality in which please show us the working of insert-step, browser console and other functions as well.

**Comment by Divy123** [comment]
Also, can you please give reference to the debugging feature of vscode.

I personally use vscode, but let's see other opinions as well like Varun suggested.

**Comment by HarshKhandeparkar** [comment]
Gift? I didn't understand fully. Could you please elaborate. Ty.

**Comment by HarshKhandeparkar** [comment]
Reference to vscode debugging feature? Uh. I don't know if there are docs about it anywhere else, maybe a blog. I have only seen it in the editor recently. You can also have a look at it. There will usually be a help button somewhere on the navbar, it will point you to the docs or otherwise, google is your best friend.

**Comment by Divy123** [comment]
Actually it was gif, my bad.

I will have a look at the debugger but would definitely request your thoughts upon this @tech4GT

**Comment by HarshKhandeparkar** [comment]
Very useful. Just found this.
* https://code.visualstudio.com/docs/editor/debugging
* https://code.visualstudio.com/docs/introvideos/debugging
* https://code.visualstudio.com/docs/nodejs/nodejs-debugging

**Comment by Divy123** [comment]
Thanks a lot and I request you to provide the GIF a bit fast as I am also working on the same so if this fixes, I can have a look at other parts of the code.

**Comment by HarshKhandeparkar** [comment]
Oh ok GIF. Lol. Happens many times. Were you able to test it out locally? I
will send the GIFs etc as soon as I get time.

Another thing, as soon as this scopeQuery PR gets merged, I will start work
on migrating to yarn which won't take very long. That can be done using a
few commands.

Thanks a lot!

**Comment by HarshKhandeparkar** [comment]
If you are able to check it out locally, it would be better I guess as there is not guarantee that I will get time soon. I might be able to do a little work today though, if time permits. I will try my best.

**Comment by Divy123** [comment]
Sure, cool!!

**Comment by Divy123** [comment]
@jywarren currently there are many callback hell in our code base. I highly recommend shifting to promises which not only leads to more reliable code but better readable one too.

**Comment by jywarren** [comment]
That would be cool!
